packages/protocol/contracts/L2/DelegateOwner.sol

```diff
@@ -69,22 +79,23 @@ contract DelegateOwner is EssentialContract, IMessageInvocable {
         payable
         onlyFromNamed(LibStrings.B_BRIDGE)
     {
-        (uint64 txId, address target, bytes memory txdata) =
-            abi.decode(_data, (uint64, address, bytes));
+        (uint64 txId, address target, bool isDelegateCall, bool requireSuccess, bytes memory txdata)
+        = abi.decode(_data, (uint64, address, bool, bool, bytes));
 
-        if (txId != nextTxId) revert DO_INVALID_TX_ID();
+        if (txId != nextTxId++) revert DO_INVALID_TX_ID();
 
         IBridge.Context memory ctx = IBridge(msg.sender).context();
         if (ctx.srcChainId != l1ChainId || ctx.from != realOwner) {
             revert DO_PERMISSION_DENIED();
         }
-        nextTxId++;
-        // Sending ether along with the function call. Although this is sending Ether from this
-        // contract back to itself, txData's function can now be payable.
-        (bool success,) = target.call{ value: msg.value }(txdata);
-        if (!success) revert DO_TX_REVERTED();
 
-        emit TransactionExecuted(txId, target, bytes4(txdata));
+        (bool success, bytes memory returnData) =
+            isDelegateCall ? target.delegatecall(txdata) : target.call{ value: msg.value }(txdata);
+
+        if (requireSuccess && !success) revert DO_TX_REVERTED();
```

Comment:
Since it can be a lot (and complex) revert reason, simply reverting with DO_TX_REVERTED() would make the debugging horrendous.

What about a revert like this:
`revert(_getRevertMsg(result));`

Where an example `_getRevertMsg()` function implementation can be found here: https://github.com/boringcrypto/BoringSolidity/blob/78f4817d9c0d95fe9c45cd42e307ccd22cf5f4fc/contracts/BoringBatchable.sol#L19